### PR TITLE
rustfmt 1.4.38 (new formula)

### PR DESCRIPTION
**Pull request by jimexist**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/rustfmt.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+class Rustfmt < Formula
+  desc "Format Rust code"
+  homepage "https://rust-lang.github.io/rustfmt/"
+  url "https://github.com/rust-lang/rustfmt/archive/refs/tags/v1.4.38.tar.gz"
+  sha256 "4856cac540a6d894bc019285b690cebd2076aa8067bd8509ec176645c47a31bf"
+  license any_of: ["MIT", "Apache-2.0"]
+  head "https://github.com/rust-lang/rustfmt.git", branch: "master"
+
+  depends_on "rust" => [:build, :test]
+  depends_on "rustup-init" => :build
```
Comment (carlocab) [inline Formula/rustfmt.rb:10]: Do we need `rustup` here? Does using `cargo` from the `rust` formula not work?

Comment (Jimexist) [reply Formula/rustfmt.rb:10]: we need `rustup` in order to use nightly toolchain which is required to build this formula from source

Comment (SMillerDev) [reply Formula/rustfmt.rb:10]: Is there no way to build it with a stable toolchain?

Comment (Jimexist) [reply Formula/rustfmt.rb:10]: It is documented so in the project readme and I can reproduce the error while building locally without nightly toolchain 

Any suggestion is welcome

Comment (Jimexist) [reply Formula/rustfmt.rb:10]: [link to the readme][1]

[1]: https://github.com/rust-lang/rustfmt/tree/v1.4.38#installing-from-source

Comment (Jimexist) [reply Formula/rustfmt.rb:10]: local reproduction:

```
rustfmt on  HEAD (efa8f55) is 📦 v1.4.38 via 🦀 nightly-2021-10-20-x86_64-apple-darwin
❯ cargo b
   Compiling rustfmt-nightly v1.4.38 (/Users/jiayu_liu/GitHub/rustfmt)
error[E0463]: can't find crate for `rustc_ast`
  --> src/lib.rs:16:1
   |
16 | extern crate rustc_ast;
   | ^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

For more information about this error, try `rustc --explain E0463`.
error: could not compile `rustfmt-nightly` due to previous error
```

`rustc_ast` is nightly only

File: Formula/rustfmt.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  def install
+    system "#{Formula["rustup-init"].bin}/rustup-init", "-qy", "--no-modify-path"
+    ENV.prepend_path "PATH", HOMEBREW_CACHE/"cargo_cache/bin"
+    system "rustup", "toolchain", "install", "nightly"
+    system "rustup", "component", "add", "rust-src", "rustc-dev", "llvm-tools-preview", "--toolchain", "nightly"
```
Comment (carlocab) [inline Formula/rustfmt.rb:16]: We should pin these to a date, at least, so we don't get a different build each time this is run.

File: Formula/rustfmt.rb (hunk 1 above)
Comment (carlocab) [inline Formula/rustfmt.rb:10]:
```suggestion
  depends_on "rustup-init" => :build
  depends_on "rust" => :test
```
Does this work? It doesn't seem like you should need `rust` at runtime given that you're installing everything else with `rustup`.

File: Formula/rustfmt.rb (hunk 2 above)
Comment (Jimexist) [reply Formula/rustfmt.rb]: updated

File: Formula/rustfmt.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  depends_on "rustup-init" => :build
+  depends_on "rust" => :test
+
+  def install
+    system "#{Formula["rustup-init"].bin}/rustup-init", "-qy", "--no-modify-path"
+    ENV.prepend_path "PATH", HOMEBREW_CACHE/"cargo_cache/bin"
+    # we are using nightly because rustfmt requires nightly in order to build from source
+    # pinning to nightly-2021-11-08 to avoid inconstency
+    system "rustup", "toolchain", "install", "nightly-2021-11-08"
```
Comment (carlocab) [inline Formula/rustfmt.rb:17]: Let's extract `nightly-2021-11-08` into a constant at the start of the `install` method so it's easier to change.

File: Formula/rustfmt.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    system "rustup", "component", "add", "rust-src", "rustc-dev", "llvm-tools-preview", \
+      "--toolchain", "nightly-2021-11-08"
```
Comment (carlocab) [inline Formula/rustfmt.rb:19]:
```suggestion
    system "rustup", "component", "add", "rust-src", "rustc-dev", "llvm-tools-preview",
           "--toolchain", "nightly-2021-11-08"
```
Let's align the arguments to `system`, and we don't need the `\`.

**Review by carlocab — COMMENTED** [review]
Just two more small suggestions and this should be ready to merge (in my opinion).

Also, when you make these changes, could you amend your commits to align with our style guide? Here's a summary:

Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.

For new formulae:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

For existing formulae:

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

**Comment by Jimexist** [comment]
> Just two more small suggestions and this should be ready to merge (in my opinion).
> 
> Also, when you make these changes, could you amend your commits to align with our style guide? Here's a summary:
> 
> Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.
> 
> For new formulae:
> 
> > At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.
> 
> For existing formulae:
> 
> > The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.
> 
> Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:
> 
> > * One formula per commit; one commit per formula.
> > * Keep merge commits out of the pull request.

thanks for the detailed advice. i have updated the comment and pull request. PTAL

**Review by carlocab — APPROVED** [review]
Great; thanks for your work here, @Jimexist!

**Pull request closed without merging**